library/common/http/dispatcher.cc

```diff
@@ -49,75 +71,137 @@ void Dispatcher::DirectStreamCallbacks::onHeaders(HeaderMapPtr&& headers, bool e
+  // closeRemote, in the Envoy Main thread. Hence DirectStream destruction which is posted on the
+  // Envoy Main thread's event loop will strictly happen after this direct_stream_ reference is
+  // used. @see Dispatcher::DirectStreamCallbacks::closeRemote() for more details.
+  if (direct_stream_.dispatchable(true)) {
+    ENVOY_LOG(debug, "[S{}] dispatching to platform remote reset stream",
+              direct_stream_.stream_handle_);
+    bridge_callbacks_.on_error({code, message}, bridge_callbacks_.context);
+
+    // All the terminal callbacks only cleanup if they are dispatchable.
+    // This ensures that cleanup will happen exactly one time.
+    http_dispatcher_.cleanup(direct_stream_.stream_handle_);
+  }
 }
 
-Dispatcher::DirectStream::DirectStream(envoy_stream_t stream_handle,
-                                       AsyncClient::Stream& underlying_stream,
-                                       DirectStreamCallbacksPtr&& callbacks)
-    : stream_handle_(stream_handle), underlying_stream_(underlying_stream),
-      callbacks_(std::move(callbacks)) {}
-
-AsyncClient::StreamOptions
-Dispatcher::DirectStream::toNativeStreamOptions(envoy_stream_options stream_options) {
-  AsyncClient::StreamOptions native_stream_options;
-  native_stream_options.setBufferBodyForRetry(stream_options.buffer_body_for_retry);
-  return native_stream_options;
+void Dispatcher::DirectStreamCallbacks::onCancel() {
+  // This call is guarded at the call-site @see Dispatcher::DirectStream::resetStream().
+  // Therefore, it is dispatched here without protection.
+  ENVOY_LOG(debug, "[S{}] dispatching to platform cancel stream", direct_stream_.stream_handle_);
+  bridge_callbacks_.on_cancel(bridge_callbacks_.context);
 }
 
-void Dispatcher::ready(Event::Dispatcher& event_dispatcher,
-                       Upstream::ClusterManager& cluster_manager) {
+Dispatcher::DirectStream::DirectStream(envoy_stream_t stream_handle, Dispatcher& http_dispatcher)
+    : stream_handle_(stream_handle), parent_(http_dispatcher) {}
+
+Dispatcher::DirectStream::~DirectStream() {
+  ENVOY_LOG(debug, "[S{}] destroy stream", stream_handle_);
+}
+
+// TODO(junr03): map from StreamResetReason to Envoy Mobile's error types. Right now all resets
+// will be ENVOY_STREAM_RESET.
+void Dispatcher::DirectStream::resetStream(StreamResetReason) { callbacks_->onReset(); }
+
+void Dispatcher::DirectStream::closeLocal(bool end_stream) {
+  // TODO: potentially guard against double local closure.
+  local_closed_ = end_stream;
+
+  // No cleanup happens here because cleanup always happens on remote closure or local reset.
+  // @see Dispatcher::DirectStreamCallbacks::closeRemote, and @see Dispatcher::resetStream,
+  // respectively.
+}
+
+bool Dispatcher::DirectStream::dispatchable(bool close) {
+  if (close) {
+    // Set closed to true and return true if not previously closed.
+    return !closed_.exchange(close);
+  }
+  return !closed_.load();
```

Comment:
There's a bug here, and it relates to the fact that the previous platform-level implementation checked this state (in the cases it mattered) from a serialized post-dispatch context. Here, it's necessarily the pre-dispatch context, and as it stands, cancellation could race another callback (resulting in, e.g., an `on_headers` callback being dispatched after an `on_cancel`). I expect this to be a fairly rare race, and so I'd be okay rolling forward with a FIXME, but as for the fix, I can't think of anything we can do besides having a lock that a thread holds while checking this state and performing platform dispatch.

Reply:
@goaway hmmmm actually don't we have the same problem today? Because the thread that cancel checks the atomic is _not_ the same as the thread the dispatch queue context runs in, right?

Could that be the race that @rebello95 was seeing?

Reply:
https://github.com/lyft/envoy-mobile/pull/616/commits/f3d8ff27efe6035b122e65b35ae4bdd23c44b8d6 captures how the race is solved for now. #647 opened to explore alternative solutions in the future.